vm/src/compile.rs

```diff
@@ -533,6 +533,65 @@ impl Compiler {
         self.emit(Instruction::BinaryOperation { op: i });
     }
 
+    fn compile_test(
+        &mut self,
+        expression: &ast::Expression,
+        true_label: Option<Label>,
+        false_label: Option<Label>,
+        context: EvalContext,
+    ) {
+        // Compile expression for test, and jump to label if false
+        match expression {
+            ast::Expression::BoolOp { a, op, b } => match op {
+                ast::BooleanOperator::And => {
+                    let f = false_label.unwrap_or_else(|| self.new_label());
+                    self.compile_test(a, None, Some(f), context);
+                    self.compile_test(b, true_label, false_label, context);
+                    if let None = false_label {
+                        self.set_label(f);
+                    }
+                }
+                ast::BooleanOperator::Or => {
+                    let t = true_label.unwrap_or_else(|| self.new_label());
+                    self.compile_test(a, Some(t), None, context);
+                    self.compile_test(b, true_label, false_label, context);
+                    if let None = true_label {
+                        self.set_label(t);
+                    }
+                }
+            },
+            _ => {
+                self.compile_expression(expression);
+                match context {
+                    EvalContext::Statement => {
+                        if let Some(true_label) = true_label {
+                            self.emit(Instruction::JumpIf { target: true_label });
+                        }
+                        if let Some(false_label) = false_label {
+                            self.emit(Instruction::JumpIfFalse {
+                                target: false_label,
+                            });
+                        }
+                    }
+                    EvalContext::Expression => {
+                        if let Some(true_label) = true_label {
+                            self.emit(Instruction::Duplicate);
```

Comment:
Wauw, I think I am starting to understand your point with the context. A nice example is this python code:

```python
a = 1 or False
assert type(a) is int
```
This was surprising to me!

Nice job!

Reply:
I still don't have it quite right! We leak values on the stack in cases like this. (The 1 gets left on the stack.)

```
1 and True or False
```

I think I know the fix.